### ISSUE TITLE
missing/wrong info

### ISSUE THREAD

**Issue by schriftgestalt** [googlei18n/glyphsLib#5]
I added some to the item that where mentioned in the Nodes section to be missing from the .ufo that is written from Glyphs. 
- I had implemented the storing of 'component has automatic alignment disabled' but was not sure about it. This time I found a way to prevent wrong assignment of the flag if the components where changed in another app.
- Guide.locked is stored and read correctly.
- I’m not sure if it is useful to store hints in the lib. Any change to the outlines would make it useless if not damaging the file.
- Glyph timestamps is stored and read correctly (using `com.schriftgestaltung.Glyphs.lastChange`)

And I fixed the wrong keys in fontInfo

**Comment by behdad** [comment]
+1!

Georg: I was hoping that long-term we can hand this module to you.  Would you be interested in that?

**Comment by schriftgestalt** [comment]
I was surprised to find something like this by accident. I’m find with managing it.

**Comment by behdad** [comment]
> I was surprised to find something like this by accident. I’m fine with managing it.

Huh.  I thought we announce it.  But you are right, we never did!  I'm sending a note to https://groups.google.com/forum/#!forum/fonttools now.  It took months to get all the release approvals for it, so by the time we got it out we were just happy that it happened.

**Comment by behdad** [comment]
> I’m fine with managing it.

Great.  Added you as collaborator.  @jamesgk wrote most of this.  Happy to let the two of you maintain it!

**Comment by schriftgestalt** [comment]
When I’m at it. I have a few questions:
- There is some "redundant" data in the font font:
  - `openTypeNamePreferredFamilyName`: Do we need this if it is the same as the `familyName`? (same for `styleMapFamilyName`)
  - `openTypeNameUniqueID`: Shouldn’t it be set only if different from `version;PostScritName`. And halfway decent export script should be able to set then from the default.

**Comment by twardoch** [comment]
This is an interesting question. In the SFNT name table, the Typographic Family Name (TFN, in the past called Preferred Family Name) field is supposed to be empty if the regular Family Name (in FontLab called Styling Group Name, SGN) is the same. 

But that's a specific "compression" aspect of SFNT. In UFO, TFN and SGN are likely to be edited independently. UFO os, after all, not necessarily an abstracted convenience representation of SFNT, but an abstracted digital representation of a typeface. UFO is of course influenced by SFNT but it should be only if that makes sense, and should not sacrifice the elegance. 

The TFN field in UFO is cslled openTypePreferredFamilyName but to me this is largely for historical reasons. Conceptually, this field represents the common linking element of all typgraphic styles in a family, andvI would find it weird if it ever was empty. 

In other words: the handling of the name table in SFNT had been messy in the last 20 years. Higher-level formats like UFO should try to clean some of this mess up instead of carrying on the baggage of trash.

**Comment by twardoch** [comment]
In other words: SFNT's main purpose is for efficient READING. SFNT does not need to be elegant for authoring purposes (though it wouldn't harm), which is why, despite some reservations, we put up with stuff like the CFF table. In a way, SFNT is a read-only format. 

But UFO's main purpose is authoring. There, we need to consider redundancies of storage vs. redundancies of operations. The TFN should only be empty in SFNT if it's identical to SGN. But there are situations where you want to keep all fonts in the same Typographic Family but rearrange the Styling Groups (StyleMaps) around. If you change the SGN, then you'd have to remember to put the old SGN into the TFN if it's empty. That's redundancies of operations. 

In a development format, some redundancies are useful. When authoring, people and tools make mistakes, or they only do partial edits, or cascading, non-concurrent edits. 

For example, in development formats, it's useful to have both Unicodes and glyphnames because you may keep one while changing the other and vice-versa. Yes, this leads to potential redundancy of both data and operations, but that's to some extent safeguard redundancy. (Though I don't think putting the Unicode codepoint into the .glif file makes much sense.) 

In optimized SFNT, glyphnames are redundant, because SFNT is meant to be efficient for machine processing.

**Comment by schriftgestalt** [comment]
> But there are situations where you want to keep all fonts in the same Typographic Family but rearrange the Styling Groups (StyleMaps) around.

But then you can add both fields. In all other cases it is more work to always have to change the same data several times. If someone puts thinks it is a good idea to duplicate the names, fine. But the tools should not to that on their own by default.

**Comment by behdad** [comment]
I agree with Georg that duplication is undesirable from a maintenance point of view :).

**Comment by jamesgk** [comment]
Hi @schriftgestalt!

> `openTypeNamePreferredFamilyName`: Do we need this if it is the same as the `familyName`? (same for `styleMapFamilyName`)

Sounds like the conclusion here is, no we don't need them?

> `openTypeNameUniqueID`: Shouldn’t it be set only if different from `version;PostScritName`. And halfway decent export script should be able to set then from the default.

That's fair. I guess my thinking generally (also applies to family name stuff) was that this library should generate UFOs with as much info filled out as possible (a "fully functional" UFO rather than a UFO as an intermediate format). But filling out as little as possible is arguably a better approach (UFOs largely function as a intermediate format anyways, huh).

> How to pick the "style" info when exporting an master? The master weight/width setting (of the master) do not actually mean to be accurate.

Hmm. Would it be better to use weightValue/widthValue? Or something else?

> I’m not sure if it is useful to store hints in the lib. Any change to the outlines would make it useless if not damaging the file.

I wanted to include as much Glyphs data as possible, for round-trip compatibility. But I guess there isn't really a use-case for loading externally-generated Glyphs hints from a UFO into the Glyphs app. So yea I'm fine with excluding this.

> And I fixed the wrong keys in fontInfo

Ack, there are actually more of these keys, and I forgot to update the readme. The code that handles this is [here](https://github.com/googlei18n/glyphs2ufo/blob/master/Lib/glyphs2ufo/torf.py#L227) (starting with the `opentype_attr_prefix_pairs` assignment), basically they're all keys which need "openType[table]" or "postscript" prepended. Spec is [here](http://unifiedfontobject.org/versions/ufo3/fontinfo.html) for reference.

**Comment by jamesgk** [comment]
Also, I'm thinking glyphs2ufo should output glyph group names according to the ufo3 spec, following the discussion at https://github.com/behdad/fonttools/commit/37a79a330c83a5fbc793aec94028b525f2f30fb2#commitcomment-14101931.

So basically a name like "@MMK_L_A" would be changed to "public.kern1.A". Is this something you'd be interested in doing in Glyphs itself? I'm not completely sure how glyph group naming works in Glyphs now.

**Comment by schriftgestalt** [comment]
In Glyphs I still use the "@MMK_L_A" syntax. I though about switching to "public.kern1.A".

**Comment by schriftgestalt** [comment]
>> How to pick the "style" info when exporting an master? The master weight/width setting (of the master) do not actually mean to be accurate.

> Hmm. Would it be better to use weightValue/widthValue? Or something else?

The problem is to decide what weight/width the master represents. The weight/width class setting is only available in the instances.

**Comment by jamesgk** [comment]
> The problem is to decide what weight/width the master represents. The weight/width class setting is only available in the instances.

Ah, right. Well ignoring the weight/width properties of masters, I can only see the option of calculating master weight/width as fractional values mapped from the range of instance weight/widths, then naming them after either the instance or OpenType spec weight/width class they're closest to, e.g. 0.1 for Thin, 0.3 for Light.... But this seems tricky (have to avoid two masters sharing the same names).

Anyways, why not trust the master weight/width properties? As far as I can tell they are used in the Glyphs UI, and would be the expected values for the master names.

**Comment by schriftgestalt** [comment]
> Anyways, why not trust the master weight/width properties? As far as I can tell they are used in the Glyphs UI, and would be the expected values for the master names.

You are right.

**Comment by schriftgestalt** [comment]
There is one issue with this. For one master/one instance files, it would make sense to use the weight settings from the instance as it has more options.

**Comment by jamesgk** [comment]
> There is one issue with this. For one master/one instance files, it would make sense to use the weight settings from the instance as it has more options.

I don't understand. What do you mean by "more options"?

**Comment by schriftgestalt** [comment]
This is about exporting the masters of a file not instances. 
In Glyphs, masters and instances have there own weigh and width settings - for totally different purposes. In the masters it is only about icons in the toolbar. The style name of the instances is what goes in the font. So in a one master one instance setting, it is pretty clear. But if you have more masters and/or instances it is not as easy. Because instances are usually interpolated you can’t just pick a instance to take a style name from. Then you need to use the setting in the master.

**Comment by brawer** [comment]
Are there any remaining tasks on this bug? If not, can it be closed?

**Comment by jamesgk** [comment]
I think it should be closed

**Comment by brawer** [comment]
OK, closing this bug then.

**Comment by khaledhosny** [comment]
AFAIK, feaLib will not set mark class unless the glyph has a mark anchor, but it is not a requirement for mark glyph to always have an anchor. IMHO, setting glyph classes explicitly should be better if they are already known.

**Comment by anthrotype** [comment]
@khaledhosny I think you were commenting on the other issue about GlyphClassDef?
https://github.com/googlei18n/glyphsLib/issues/85#issuecomment-274524660

**Comment by khaledhosny** [comment]
Right, sorry. Opened the wrong link I guess. Moved my comment to the right place (feel free to delete the ones here if you want to).
